Source/Bogus/Extensions/Denmark/ExtensionsForDenmark.cs

```diff
@@ -8,18 +11,168 @@ public static class ExtensionsForDenmark
    /// <summary>
    /// Danish Personal Identification number
    /// </summary>
-   public static string Cpr(this Person p)
+   /// <param name="p">The holder.</param>
+   /// <param name="validChecksum">
+   ///   Indicates whether the generated CPR number should have a valid checksum or not.
+   /// </param>
+   public static string Cpr(this Person p, bool validChecksum = true)
    {
       const string Key = nameof(ExtensionsForDenmark) + "CPR";
-      if( p.context.ContainsKey(Key) )
+      if (p.context.ContainsKey(Key))
       {
          return p.context[Key] as string;
       }
 
+      /*
+          DDMMYY-XXXX
+          | | |  |  
+          | | |  |
+          | | |  |
+          | | |  |----> (X)Individual number
+          | | |-------> (Y)Year (last two digits)
+          | |---------> (M)Month
+          |-----------> (D)Day
+
+          The individual number has to be even for women and odd for men.
+
+          As of 2007 there is no longer a requirement for a checksum with a modulo algorithm.
+      
+         https://cpr.dk/cpr-systemet/opbygning-af-cpr-nummeret
+            
+         https://da.wikipedia.org/wiki/CPR-nummer
+               
+         https://www.cprgenerator.net/metode
+      */
+
       var r = p.Random;
-      var final = $"{p.DateOfBirth:ddMMyy}-{r.Replace("####")}";
+      string birthDate = $"{p.DateOfBirth:ddMMyy}";
+      string individualNumber;
+      string checksum;
+      bool hasValidChecksum;
+
+      if (validChecksum)
+      {
+         do
+         {
+            individualNumber = GenerateIndividualThreeDigitNumber(r, p.DateOfBirth.Year);
+            hasValidChecksum = GenerateChecksum(birthDate, p.Gender, individualNumber, out checksum);
+         } while (!hasValidChecksum);
+      }
+      else
+      {
+         checksum = string.Empty;
+         individualNumber = GenerateIndividualFourDigitNumber(r, p.Gender, p.DateOfBirth.Year);
+      }
+
+      var final = $"{birthDate}-{individualNumber}{checksum}";
 
       p.context[Key] = final;
       return final;
    }
+
+   private static string GenerateIndividualFourDigitNumber(Randomizer r, DataSets.Name.Gender gender, int year)
+   {
+      int from;
+      int to;
+
+      if (1858 <= year && year <= 1899)
+      {
+         from = 5000;
+         to = 8999;
+      }
+      else if (1900 <= year && year <= 1936)
+      {
+         from = 0;
+         to = 3999;
+      }
+      else if (1937 <= year && year <= 1999)
+      {
+         from = 0;
+         to = 4999;
+      }
+      else if (2000 <= year && year <= 2036)
+      {
+         from = 4000;
+         to = 9999;
+      }
+      else if (2037 <= year && year <= 2057)
+      {
+         from = 5000;
+         to = 9999;
+      }
+      else
+      {
+         throw new ArgumentOutOfRangeException(nameof(year), $"{nameof(year)} must be between 1854 and 2039.");
```

Comment:
The else if above suggest the range ends at 2057, which year is correct?

Reply:
This should be fixed in the latest commit. Please have a look.